activesupport/lib/active_support/execution_context.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module ExecutionContext # :nodoc:
+    @after_change_callbacks = []
+    class << self
+      def after_change(&block)
+        @after_change_callbacks << block
+      end
+
+      # Updates the execution context. If a block is given, it resets the provided keys to their
+      # previous value once the block exits.
+      def set(**options)
+        options.symbolize_keys!
+        keys = options.keys
+
+        store = self.store
+
+        previous_context = keys.zip(store.values_at(*keys)).to_h
+
+        store.merge!(options)
+        @after_change_callbacks.each(&:call)
+
+        if block_given?
+          begin
+            yield
+          ensure
+            store.merge!(previous_context)
+            @after_change_callbacks.each(&:call)
+          end
+        end
+      end
+
+      def [](key)
+        store[key.to_sym]
+      end
```

Comment:
I'm also tempted to remove `#[]` so that it can't be abused as some kind of "global variable".

Reply:
Maybe we could document that it's not meant to be used as such, and that would be clear enough? So it's technically considered private API.

I like how `[]=` reads in the call sites, versus the blocks because we aren't necessarily wanting this to be for one off blocks but having state assigned that's auto-cleared by `Rails.application.executor` so it's a hook into that lifecycle. I think the intent is clearer.

Reply:
Well, I was suggesting to remove `#[]`, not `#[]=`.

Reply:
Oh right, 👍 on removing that.

Reply:
Sounds good. I'll do that tomorrow and then merge.

If you are doing review and have the time for it, could I pick your interest on https://github.com/rails/rails/pull/43596 ?